Migrate HTML media/video element logs to efficient version

pvollan opened the pull request:
#### 132f2ebbb5fba4a13d7f873cdcc08d33b4fae2f8
<pre>
Migrate HTML media/video element logs to efficient version
<a href="https://bugs.webkit.org/show_bug.cgi?id=294648">https://bugs.webkit.org/show_bug.cgi?id=294648</a>
<a href="https://rdar.apple.com/153696898">rdar://153696898</a>

Reviewed by Eric Carlson.

Migrate HTML media/video element logs to efficient version for log forwarding.
This patch also addresses an issue where logs are emitted in a private browsing
context.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::prepareForLoad):
(WebCore::HTMLMediaElement::selectMediaResource):
(WebCore::HTMLMediaElement::setNetworkState):
(WebCore::HTMLMediaElement::setReadyState):
(WebCore::HTMLMediaElement::setPlaybackRate):
(WebCore::HTMLMediaElement::pause):
(WebCore::HTMLMediaElement::pauseInternal):
(WebCore::HTMLMediaElement::setVolume):
(WebCore::HTMLMediaElement::setMutedInternal):
(WebCore::HTMLMediaElement::removeAudioTrack):
(WebCore::HTMLMediaElement::scheduleConfigureTextTracks):
(WebCore::HTMLMediaElement::mediaPlayerDurationChanged):
(WebCore::HTMLMediaElement::mediaPlayerSizeChanged):
(WebCore::HTMLMediaElement::scheduleMediaEngineWasUpdated):
(WebCore::HTMLMediaElement::mediaPlayerCharacteristicChanged):
(WebCore::HTMLMediaElement::setAutoplayEventPlaybackState):
(WebCore::HTMLMediaElement::sceneIdentifierDidChange):
(WebCore::HTMLMediaElement::setShowPosterFlag):
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::scheduleResizeEvent):
(WebCore::HTMLVideoElement::mediaPlayerFirstVideoFrameAvailable):
* Source/WebCore/platform/LogMessages.in:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::mediaPlayerRenderingModeChanged):
(WebCore::HTMLVideoElement::scheduleResizeEvent):
(WebCore::HTMLVideoElement::mediaPlayerFirstVideoFrameAvailable):

Canonical link: <a href="https://commits.webkit.org/297325@main">https://commits.webkit.org/297325@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8be3669478cae6a442b837d51732099bb2c706e5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111284 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21415 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117316 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61552 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113246 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31631 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39532 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84584 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25260 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100192 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65030 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18333 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61135 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94641 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18400 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120455 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38333 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28486 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93509 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38709 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96467 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93333 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38432 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16207 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34330 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17949 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38222 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43699 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37887 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41220 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39589 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/html/HTMLVideoElement.cpp:
```diff
@@ -319,7 +319,7 @@ bool HTMLVideoElement::shouldDisplayPosterImage() const
 void HTMLVideoElement::mediaPlayerFirstVideoFrameAvailable()
 {
-    ALWAYS_LOG(LOGIDENTIFIER, "m_showPoster = ", showPosterFlag());
+    HTMLVIDEOELEMENT_RELEASE_LOG(HTMLVIDEOELEMENT_MEDIAPLAYERFIRSTVIDEOFRAMEAVAILABLE, showPosterFlag());
```
This is rather verbose, is there a way we could make things shorter somehow?
Can we use `__FILE__` and `__func__`?

Also we have twice the `HTMLVIDEOELEMENT_`, maybe we could have it only once?

pvollan replied:
That is a good point! I will look closer into this.

Thanks for reviewing!

pvollan replied:
In the latest patch, I have removed the `HTMLVIDEOELEMENT_` part of the format string identifier.

It may not be feasible to use `__func__` in this context, since it cannot be used in a log format string because it is not a string literal. The improved efficiency in this patch is based on having as much as possible of the logs in the log format strings, since the format string is not sent over IPC.

Thanks for reviewing!

youennf commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -9441,7 +9440,7 @@ void HTMLMediaElement::visibilityAdjustmentStateDidChange()
 {
     if (RefPtr page = document().page()) {
-        ALWAYS_LOG(LOGIDENTIFIER, page->sceneIdentifier());
+        HTMLMEDIAELEMENT_RELEASE_LOG(HTMLMEDIAELEMENT_SCENEIDENTIFIERDIDCHANGE, page->sceneIdentifier().utf8().data());
```
It is somehow unfortunate to have to call utf8 here?
Ideally, we would do it like ALWAYS_LOG, not sure whether feasible.

pvollan replied:
ALWAYS_LOG is also converting all log arguments to String and utf8:

```
    template<typename... Argument>
    static inline void logVerbose(WTFLogChannel& channel, WTFLogLevel level, const char* file, const char* function, int line, const Argument&... arguments)
    {
        auto logMessage = makeString(LogArgument<Argument>::toString(arguments)...);
        ...
        SUPPRESS_UNRETAINED_LOCAL os_log(channel.osLogChannel, "%{public}s", logMessage.utf8().data());
```

eric-carlson replied:
`ALWAYS_LOG` and family only call `.utf8().data()` on the composed log string, which is created with `makeString`, so String arguments are not converted.

pvollan replied:
Indirectly, String arguments will be converted, since they are substrings of the String returned by `makeString`, right? Or perhaps there is something I am missing here?

Thanks for reviewing!

youennf commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -3194,7 +3193,7 @@ void HTMLMediaElement::setReadyState(MediaPlayer::ReadyState state)
     m_tracksAreReady = tracksAreReady;
 
-    ALWAYS_LOG(LOGIDENTIFIER, "new state = ", state, ", current state = ", m_readyState);
+    HTMLMEDIAELEMENT_RELEASE_LOG(HTMLMEDIAELEMENT_SETREADYSTATE, convertEnumerationToString(state).utf8().data(), convertEnumerationToString(m_readyState).utf8().data());
```
Ideally, we would do the conversion if we are actually logging.
`ALWAYS_LOG  / Logger::logAlwaysVerbose` will do the conversion only if `willLog` will return true.
We are losing some efficiency here.

pvollan replied:
That is a good point! This has been resolved in the lates patch by adding a call to `willLog` in `HTMLMEDIAELEMENT_RELEASE_LOG`.

youennf reviewed (commented):
Is there a possibility to update `LoggerHelper` `log`, `logVerbose`... implementation to fix the potential inefficiencies?

pvollan committed:
Migrate HTML media/video element logs to efficient version
https://bugs.webkit.org/show_bug.cgi?id=294648
rdar://153696898

Reviewed by Eric Carlson.

Migrate HTML media/video element logs to efficient version for log forwarding.
This patch also addresses an issue where logs are emitted in a private browsing
context.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::prepareForLoad):
(WebCore::HTMLMediaElement::selectMediaResource):
(WebCore::HTMLMediaElement::setNetworkState):
(WebCore::HTMLMediaElement::setReadyState):
(WebCore::HTMLMediaElement::setPlaybackRate):
(WebCore::HTMLMediaElement::pause):
(WebCore::HTMLMediaElement::pauseInternal):
(WebCore::HTMLMediaElement::setVolume):
(WebCore::HTMLMediaElement::setMutedInternal):
(WebCore::HTMLMediaElement::removeAudioTrack):
(WebCore::HTMLMediaElement::scheduleConfigureTextTracks):
(WebCore::HTMLMediaElement::mediaPlayerDurationChanged):
(WebCore::HTMLMediaElement::mediaPlayerSizeChanged):
(WebCore::HTMLMediaElement::scheduleMediaEngineWasUpdated):
(WebCore::HTMLMediaElement::mediaPlayerCharacteristicChanged):
(WebCore::HTMLMediaElement::setAutoplayEventPlaybackState):
(WebCore::HTMLMediaElement::sceneIdentifierDidChange):
(WebCore::HTMLMediaElement::setShowPosterFlag):
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::scheduleResizeEvent):
(WebCore::HTMLVideoElement::mediaPlayerFirstVideoFrameAvailable):
* Source/WebCore/platform/LogMessages.in:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::mediaPlayerRenderingModeChanged):
(WebCore::HTMLVideoElement::scheduleResizeEvent):
(WebCore::HTMLVideoElement::mediaPlayerFirstVideoFrameAvailable):

Canonical link: https://commits.webkit.org/297325@main
